Allow refType to be defined in @prop and @arrayProp

mfulop opened the pull request:
Merged changes from pr #149, solved conflicts, added itemsRefType (works the same as refType) to @arrayProps and tests

hasezoey commented on src/prop.ts:
```diff
@@ -369,22 +387,15 @@ function baseProp(rawOptions: any, Type: any, target: any, key: string, whatis:
-export function prop(options: PropOptionsWithValidate = {}) {
-  return (target: any, key: string) => {
-    const Type = (Reflect as any).getMetadata('design:type', target, key);
+export const prop = (options: PropOptionsWithValidate = {}) => (target: any, key: string) => {
```
please use the new style of functions

hasezoey commented on test/models/refTestBuffer.ts:
```diff
@@ -0,0 +1,11 @@
+
+import { prop, Typegoose } from '../../src/typegoose';
+
+export class RefTestBuffer extends Typegoose {
```
could you combine these into one file?

hasezoey commented on test/models/refTestNumber.ts:
```diff
@@ -0,0 +1,11 @@
+
+import { prop, Typegoose } from '../../src/typegoose';
+
+export class RefTestNumber extends Typegoose {
```
could you combine these into one file?

hasezoey commented on test/models/refTestString.ts:
```diff
@@ -0,0 +1,11 @@
+
+import { prop, Typegoose } from '../../src/typegoose';
+
+export class RefTestString extends Typegoose {
```
could you combine these into one file?

hasezoey commented on test/tests/ref.test.ts:
```diff
@@ -0,0 +1,146 @@
+    const id1: string = 'testid1';
+    const id2: string = 'testid2';
+
+    const t = new RefTest();
```
could you give it a proper variable name?

hasezoey commented on test/tests/ref.test.ts:
```diff
@@ -0,0 +1,146 @@
+    const id2: string = 'testid2';
+
+    const t = new RefTest();
+    t.refFieldString = id1; // should not throw compilation error
```
could you remove `// should not throw compilation error` on all lines?

mfulop replied:
Yes. The idea was that these lines are actually test of the Ref declaration being correct. Otherwise test compilation would fail on these lines

hasezoey reviewed (changes requested):
when these are resolved, could you squash some of your commits?

i would recommend reverting 00ddb93 and then rebase, not merge :)

mfulop commented:
Sure will do. vscode merge somehow messed up things, will try another way, sorry

hasezoey commented:
i will review the changed in 60aa3ca tomorrow

mfulop commented:
ok, you might want to look at ab91ee0 instead

mfulop commented:
Btw though it’s working, I think Ref is now a bit too liberal. To be 100% correct, in Ref<R,T>, T should be R or the type of the designated id field of R.
Sadly it is beyond my current typescript knowledge :)

hasezoey commented on src/prop.ts:
```diff
@@ -58,6 +58,7 @@ export interface PropOptions extends BasePropOptions {
   ref?: any;
   /** Take the Path and try to resolve it to a Model */
   refPath?: string;
+  refType?: 'number' | 'string' | 'Buffer' | 'ObjectID';
```
change to lower case, just to keep everything normal
```suggestion
  refType?: 'number' | 'string' | 'buffer' | 'objectid;
```

hasezoey commented on src/prop.ts:
```diff
@@ -197,18 +198,35 @@ function baseProp(rawOptions: any, Type: any, target: any, key: string, whatis:
 
+  let refType;
+
+  switch (rawOptions.itemsRefType || rawOptions.refType) {
```
change to lowercase to keep everything normal, like above, and everything following
```suggestion
  switch ((rawOptions.itemsRefType || rawOptions.refType).toLowerCase()) {
```

hasezoey commented on src/prop.ts:
```diff
@@ -197,18 +198,35 @@ function baseProp(rawOptions: any, Type: any, target: any, key: string, whatis:
+  case 'string':
+    refType = mongoose.Schema.Types.String;
+    break;
+  case 'Buffer':
```
```suggestion
  case 'buffer':
```

hasezoey commented on src/prop.ts:
```diff
@@ -197,18 +198,35 @@ function baseProp(rawOptions: any, Type: any, target: any, key: string, whatis:
+  case 'Buffer':
+    refType = mongoose.Schema.Types.Buffer;
+    break;
+  default:
```
```suggestion
  case 'objectid':
  default:
```

hasezoey commented on src/prop.ts:
```diff
@@ -395,7 +413,10 @@ export interface ArrayPropOptions extends BasePropOptions {
   /** Same as {@link PropOptions.refPath}, only that it is for an array */
   itemsRefPath?: any;
+  /** Same as {@link PropOptions.refType}, only that it is for an array */
+  itemsRefType?: 'number' | 'string' | 'Buffer' | 'ObjectID';
```
```suggestion
  itemsRefType?: 'number' | 'string' | 'buffer' | 'objectid';
```

hasezoey reviewed (changes requested):
i would request setting the types to all lowercase to keep everything in style and not "string" and "Buffer" just to write it easy

PS: sorry that it took so long, my Desktop Environment just kept crashing

PPS: could you please squash all your commits? (`git rebase -i master`)
and try to reword your commits to comply with [CONTRIBUTING.md](https://github.com/szokodiakos/typegoose/blob/master/.github/CONTRIBUTING.md#how-to-structure-commits) (~~if not, i will at the merge squash & reword it~~)

mfulop commented:
How about now? 
Another thing came to mind before we finalize this. I merged Matt's changes from #149 regarding the refType, but does that have to be string? Would not specifying actual schema types make more sense? (it would to me...):
`@prop({ ref: Car, refType: mongoose.Schema.Types.String })`

hasezoey commented:
>  I merged Matt's changes from #149 regarding the refType, but does that have to be string?

yes, would be a better idea, only that "I merged Matt's changes" doesnt not seem to be actually merged (its still string)

mfulop commented:
OK, so here c0ab190 is how I would do it. I think it's much cleaner this way. What do you think?

hasezoey commented on src/prop.ts:
```diff
@@ -201,14 +208,14 @@ function baseProp(rawOptions: any, Type: any, target: any, key: string, whatis:
     schema[name][key] = {
       ...schema[name][key],
-      type: mongoose.Schema.Types.ObjectId,
+      type: rawOptions.refType || mongoose.Schema.Types.ObjectId,
```
i think it would be better to do something like `rawOptions.refType = rawOptions.refType ? rawOptions.refType : mongoose.Schema.Types.ObjectId` instead of repetitively doing this

hasezoey commented:
I noticed after some testing, that if i do 
```ts
const test = new RefTestModel({ refFieldString: new RefTestStringModel() }); // just as an example
if (isDocument(test.refFieldString)) {
  test.refFieldString // i mean here
}
```
the type of `test.refFieldString` in the if is `InstanceType<T> | (string & Document)` where it should only be `InstanceType<T>`

hasezoey commented:
got it fixed, please modify typeguards.ts:
```ts
export function isDocument<T, S extends RefType>(doc: Ref<T, S>): doc is InstanceType<T> {
```
and
```ts
export function isDocumentArray<T, S extends RefType>(docs: Ref<T, S>[]): docs is InstanceType<T>[] {
```

mfulop commented:
I implemented all the changes and added the test. That actually brought up this:
Shouldn't 
`export type Ref<R, T extends RefType = mongoose.Schema.Types.ObjectId> = R | T;`
be
`export type Ref<R, T extends RefType = mongoose.Types.ObjectId> = R | T;`
instead? (the actual values are not SchemaTypes but ObjectIds).

But would this change break compatibility?

hasezoey commented:
a mongoose issue explaining the difference (for further readers) https://github.com/Automattic/mongoose/issues/1671

so yes, it would be *more* right to use `Types.T` insteace of `Schema.Types.T`
and i cant see any obvious "breaking" changes, "Ref<T, S>" is just a type and so-or-so it works as expected

so i would say, yes it would be good to keep it so

mfulop commented:
This: `new RefTestModel({ refField: mongoose.Types.ObjectId() })` is actually not possible without casting when using SchemaType

hasezoey commented on test/models/refTests.ts:
```diff
@@ -0,0 +1,79 @@
+import { arrayProp, prop, Ref, Typegoose } from '../../src/typegoose';
+
+export class RefTestBuffer extends Typegoose {
+  @prop() public _id: mongoose.Schema.Types.Buffer;
```
it would be good if we could keep a consistent class style
so i ask, could you change this (and the following classes) to use the following style?
```ts
@prop()
public _id; // separate line
```

hasezoey commented on test/tests/ref.test.ts:
```diff
@@ -0,0 +1,189 @@
+  });
+
+  it('check reference with string _id', async () => {
+    const id1: string = 'testid1';
```
to keep consistent style, could you remove the "redundant" typedef?

hasezoey commented on test/tests/ref.test.ts:
```diff
@@ -0,0 +1,189 @@
+
+  it('check reference with string _id', async () => {
+    const id1: string = 'testid1';
+    const id2: string = 'testid2';
```
to keep consistent style, could you remove the "redundant" typedef?

hasezoey commented on test/tests/ref.test.ts:
```diff
@@ -0,0 +1,189 @@
+    expect(isDocumentArray(idFields.refArrayNumber)).to.equal(false);
+    expect(isDocumentArray(idFields.refArrayBuffer)).to.equal(false);
+
+    const populatedFields = new RefTestModel({
```
it would be good if this could be changed to store it and retrieve it, just to be "extra" sure
```suggestion
    const saved = await RefTestModel.create({
```

hasezoey commented:
> This: `new RefTestModel({ refField: mongoose.Types.ObjectId() })` is actually not possible without casting when using SchemaType

ok, then lets keep it at `Schema.Types.T`

mfulop commented:
Then its done ;-)

hasezoey commented on test/tests/ref.test.ts:
```diff
@@ -0,0 +1,189 @@
+      refFieldBuffer: new RefTestBufferModel({ _id: Buffer.from([1, 2, 3, 4]) }),
+      refArrayBuffer: [new RefTestBufferModel({ _id: Buffer.from([1, 2, 3, 4]) }),
+        new RefTestBufferModel({ _id: Buffer.from([1, 2, 3, 4]) })]
+    });
```
(in conjunction with the comment above)
```suggestion
    });
    
    const found = RefTestModel.findById(saved.id).populate('refField') //etc
```

PS: i know this is redundant, and will result in a long .populate "command", but just to be extra sure

hasezoey commented on test/tests/ref.test.ts:
```diff
@@ -0,0 +1,189 @@
+      refFieldNumber: new RefTestNumberModel({ _id: 1234 }),
+      refArrayNumber: [new RefTestNumberModel({ _id: 1234 }), new RefTestNumberModel({ _id: 1234 })],
+      refFieldBuffer: new RefTestBufferModel({ _id: Buffer.from([1, 2, 3, 4]) }),
+      refArrayBuffer: [new RefTestBufferModel({ _id: Buffer.from([1, 2, 3, 4]) }),
```
(i suggest this change, because vscode on save somehow formats it wrong ...)
```suggestion
      refArrayBuffer: [
        new RefTestBufferModel({ _id: Buffer.from([1, 2, 3, 4]) }),
```

hasezoey commented on test/tests/ref.test.ts:
```diff
@@ -0,0 +1,189 @@
+      refArrayNumber: [new RefTestNumberModel({ _id: 1234 }), new RefTestNumberModel({ _id: 1234 })],
+      refFieldBuffer: new RefTestBufferModel({ _id: Buffer.from([1, 2, 3, 4]) }),
+      refArrayBuffer: [new RefTestBufferModel({ _id: Buffer.from([1, 2, 3, 4]) }),
+        new RefTestBufferModel({ _id: Buffer.from([1, 2, 3, 4]) })]
```
```suggestion
        new RefTestBufferModel({ _id: Buffer.from([1, 2, 3, 4]) })
      ]
```

hasezoey reviewed (changes requested):
Everything is and looks good, just some small style changes

mfulop commented:
Could you check something for me?

I am putting together the test you requested above. When running the below snippet, the created RefTestModel has both refFieldString and refFieldNumber fields created with the corresponding ID.
However, while the RefTestNumber document is created in the database, for some reason, the RefTestString document is not. Therefore, the next populate will not find it.
What makes the difference? Or is it a race condition somewhere?

```
await RefTestModel.create({
      refFieldString: new RefTestStringModel({ _id: 'test1' }), // document not created
      refFieldNumber: new RefTestNumberModel({ _id: 1234 }), // document created
    });
```

mfulop commented:
By using create `refFieldString: await RefTestStringModel.create({ _id: 'test1' })` it works (of course) but I guess that's not the point.

hasezoey commented:
> Or is it a race condition somewhere?

typegoose nowhere sets a race condition, at least not knowingly 

---

could you push it to another branch so i can check it out?

mfulop commented:
https://github.com/mfulop/typegoose/tree/reftests/

hasezoey commented:
> However, while the RefTestNumber document is created in the database, for some reason, the RefTestString document is not.

sorry, cant confirm that, for me everything gets saved
modified it to
```ts
const { _id: populatedId } = await RefTestModel.create({
  refField: new RefTestModel(), // gets saved
  refArray: [new RefTestModel(), new RefTestModel()], // gets saved
  refFieldString: new RefTestStringModel({ _id: 'test1' }), // gets saved
  refArrayString: [
    new RefTestStringModel({ _id: 'test2' }), // gets saved
    new RefTestStringModel({ _id: 'test3' }) // gets saved
  ],
  refFieldNumber: new RefTestNumberModel({ _id: 1234 }), // gets saved
  refArrayNumber: [new RefTestNumberModel({ _id: 5678 }), new RefTestNumberModel({ _id: 9876 })], // gets saved
  refFieldBuffer: new RefTestBufferModel({ _id: Buffer.from([1, 2, 3, 4]) }), // gets saved
  refArrayBuffer: [
    new RefTestBufferModel({ _id: Buffer.from([5, 6, 7, 8]) }), // gets saved
    new RefTestBufferModel({ _id: Buffer.from([9, 8, 7, 6]) }) // gets saved
  ]
});
```

mfulop commented:
Do tests run without an error on the reftests branch for you?

mfulop commented:
Oh yes, the root document is correctly created. But do the referenced documents (eg. the one with id 5678) all exist?

hasezoey commented:
> Do tests run without an error on the reftests branch for you?

no thests do not run, because not populated properly

> Oh yes, the root document is correctly created. But do the referenced documents (eg. the one with id 5678) all exist?

i dont have any with the id with `5678`, only an array of numbers

hasezoey commented:
currently it fails at `expect(foundPopulated.refArray.length).to.equal(2);`, with `AssertionError: expected 0 to equal 2`

mfulop commented:
> currently it fails at `expect(foundPopulated.refArray.length).to.equal(2);`, with `AssertionError: expected 0 to equal 2`
yes and that means populate can't find the foreign document with that id

hasezoey commented:
> yes and that means populate can't find the foreign document with that id

i can confirm that Field `refArrayNumber` documents are saved on the DB

mfulop commented:
Ok, I will dig deeper tomorrow to see what’s going on. If the documents are there and the ids are there then populate should work.

hasezoey commented:
got it further with
```ts
const pop = Object.keys(RefTestModel.schema.obj).join();
const foundPopulated = await RefTestModel.findById(populatedId).populate(pop);
```

new error: `AssertionError: expected false to equal true` on line
```ts
expect(isDocument(foundPopulated.refField)).to.equal(true);
```

will investigate further

mfulop commented:
First thing I will try in the morning is the same with native mongoose.

hasezoey commented:
Update(4):
i noticed after disabling all other tests, that the documents were never created (including the ones i said are existing), because they have duplicate ids (like the number `5678` is used 3 times)

TL;DR: what i said was wrong, no new refs are actually saved

mfulop commented:
OK, so it is not supposed to work. Then await create() will do for tests, give me a minute

Commit:
Adding @prop({ refType }) and @arrayProp({ itemsRefType }) (fixes #181)
- added @prop({ refType })
- added @arrayProp({ itemsRefType })
- added Ref<R, T> to define Ref types with string, number of buffer id
- Use mongoose.Schema.Types as refType
- changed Ref<> value to mongoose.Types.ObjectId
- added refType support to IsDocument and isDocumentArray typeguards
- added tests

hasezoey commented on test/tests/ref.test.ts:
```diff
@@ -0,0 +1,210 @@
+      ]
+    });
+
+    const populate = ['', 'String', 'Number', 'Buffer'].map((f) => `refField${f} refArray${f}`).join(' ');
```
```suggestion
    const populate = Object.keys(RefTestModel.schema.obj);
```

mfulop replied:
_id too?

hasezoey replied:
this dosnt include _id

hasezoey replied:
```ts
const populate = Object.keys(RefTestModel.schema.obj);
// tslint:disable:no-
console.log('pop', populate);
```
outputs an array, without `_id`

mfulop commented:
> -> we _could_ implement such things, but this would mean that we would need to redo the .create / .save functions OR do an "invisible" pre-save-hook
My 2 cents - I would not... it's a quite big difference compared to how mongoose behaves on its own, could be misleading for the ones with Mongoose experience

> --> but it is out of scope for _this_ PR
I agree

hasezoey commented:
i see no problem anymore, only resolve the last review, squash your commits, and i will approve :+1:

mfulop commented:
There you go 189a6ce

And thanks for all :)

hasezoey reviewed (approved):
Everything got resolved, everything conforms with the Coding Style, Commits are structured right :+1:

hasezoey commented:
@Ben305 (if you are back) i would want to know if this should be in release 6.0.0 or release 5.10.0?

mfulop commented:
Hi!

Sorry, I was out for a bit. Actually I would appreciate this before v6. I am planning to port some of our proprietary sw to Typegoose and this was one missing feature. (There might be more but then you’ll see a PR about them anyways :) ). Unfortunately this project is now iced for a few days but I’ll be back.

hasezoey commented:
> Unfortunately this project is now iced for a few days but I’ll be back.

which project? this pr is finished an ready to go :) (ben is out for a week)

mfulop commented:
The on where I am porting stuff to run on Typegoose :)

hasezoey commented:
Merged it for now into 6.0.0(-19)

hasezoey commented:
this can be closed because it is included in v6.0.0, and will not be included a v5.x release

@Ben305
